### PR TITLE
added weighted binary cross entropy and weighted MSE objectives

### PR DESCRIPTION
**Pull request by annashcherbina**
weighted objective functions for binary cross-entropy and mean-squared error objective functions. 
w0 and w1 are lists of weights for the negative & positive class, respectively, with w0_i and w1_i corresponding to task i in a multi-tasked model. 

The benefit of the weighted objective functions is the ability to handle cases when some tasks have a much higher degree of class imbalance then others. This functionality is different from the existing weights argument to model.train because the weights argument is assigned with w_j for datapoint j, whereas the proposed objectives would assign weights at the task rather than the datapoint level.

### CODE REVIEW

**Comment by fchollet** [comment]
> This functionality is different from the existing weights argument to model.train because the weights argument is assigned with w_j for datapoint j, whereas the proposed objectives would assign weights at the task rather than the datapoint level.

I don't get it. You can set weights for tasks (in `compile`), but also for classes (`class_weight` argument  in `fit`) and for samples, what you call data points (`sample_weight` argument in `fit`).

**Comment by annashcherbina** [comment]
Ok, I think the functionality I was trying to achieve is a combination of the loss_weights in compile() and the class_weight argument in fit(). 

It seems the loss weights assign a weight for each task. And the class_weights assign a weight for each class. How are those two combined together to determine weight of class 1 for task i vs. weight of class 1 for task j? 

I would like to handle a class like this: 
w1=[1,1,5,3] 
w0=[3,4,2,9] 

So the relative w1 & w0 for the 4 tasks of interest are different.

**Comment by AvantiShri** [comment]
It looks like @fchollet is using "task" to refer to a particular output layer, but in this context @annashcherbina is using "task" to refer to a specific index WITHIN a given output layer. The loss_weights argument appears to specify one weight per output layer, because loss_weights has the same length as self.outputs (https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L494) and for a sequential model, the length of self.outputs is 1 (https://github.com/fchollet/keras/blob/master/keras/models.py#L240). It doesn't look like there is currently a way to weight different indices within a particular output layer.

**Comment by fchollet** [comment]
I think this would be too niche a use case for the time being.

**Comment by akundaje** [comment]
Hi @fchollet. The current alternative to having index-specific weights would be to create multiple Dense layers, each containing a single neuron. Not only would this force the user to use a functional model rather than a Sequential model, but when working like a large number of tasks (~200-400, which is not uncommon when doing multitasking), it would clutter the architecture of the model and make it unwieldy to read and manipulate. All major deep learning for genomics papers till now have used a single Dense layer for the output (see: Basset, DeepSEA), and all had a different class imbalance for different output indices. Adding this functionality would go a long way towards making Keras more usable for genomics. Hopefully you reconsider :)

**Comment by fchollet** [comment]
@akundaje we avoid including support for specialized use case directly into the codebase, because there is a very long tail of such use cases and it would be too much to maintain. This functionality is self-contained and only spans a few lines of code, so it would not be an issue for you to keep it as part of your own codebase. If it's useful for others, you can make it available as part of a genomics toolbox. Then if there is enough demand, we'll consider including it.

**Comment by akundaje** [comment]
@fchollet Sure. Makes sense. Thanks for clarifying.
